pkg/ccl/changefeedccl/sink.go

```diff
@@ -587,6 +639,256 @@ func (s *bufferSink) Close() error {
+func (s *cloudStorageSink) EmitRow(
+	_ context.Context, table *sqlbase.TableDescriptor, _, value []byte, updated hlc.Timestamp,
+) error {
+	if s.files == nil {
+		return errors.New(`cannot EmitRow on a closed sink`)
+	}
+
+	// Intentionally throw away the logical part of the timestamp for bucketing.
+	key := cloudStorageSinkKey{
+		Bucket:   updated.GoTime().Truncate(s.bucketSize),
+		Topic:    table.Name,
+		SchemaID: table.Version,
+		SinkID:   s.sinkID,
+		Ext:      s.ext,
+	}
+	file := s.files[key]
+	if file == nil {
+		// We could pool the bytes.Buffers if necessary, but we'd need to be
+		// careful to bound the size of the memory held by the pool.
+		file = &bytes.Buffer{}
+		s.files[key] = file
+	}
+
+	// TODO(dan): Memory monitoring for this
+	if _, err := file.Write(value); err != nil {
+		return err
+	}
+	return s.recordDelimFn(file)
+}
+
+// EmitResolvedTimestamp implements the Sink interface.
+func (s *cloudStorageSink) EmitResolvedTimestamp(
+	ctx context.Context, payload []byte, resolved hlc.Timestamp,
+) error {
+	if s.files == nil {
+		return errors.New(`cannot EmitRow on a closed sink`)
+	}
+
+	es, err := storageccl.ExportStorageFromURI(ctx, s.base.String(), s.settings)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := es.Close(); err != nil {
+			log.Warningf(ctx, `failed to close %s, resources may have leaked: %s`, s.base.String(), err)
+		}
+	}()
+
+	// resolving some given time means that every in the _previous_ bucket is
+	// finished.
+	resolvedBucket := resolved.GoTime().Truncate(s.bucketSize).Add(-time.Nanosecond)
+	name := cloudStorageFormatBucket(resolvedBucket) + `.RESOLVED`
+	if log.V(1) {
+		log.Info(ctx, "writing ", name)
+	}
+	return es.WriteFile(ctx, name, bytes.NewReader(payload))
+}
+
+// Flush implements the Sink interface.
+func (s *cloudStorageSink) Flush(ctx context.Context, ts hlc.Timestamp) error {
+	if s.files == nil {
+		return errors.New(`cannot Flush on a closed sink`)
+	}
+
+	var gcKeys []cloudStorageSinkKey
+	for key, file := range s.files {
+		// Any files where the bucket begin is `>= ts` don't need to be flushed
+		// because of the Flush contract w.r.t. `ts`. (Bucket begin time is
+		// exclusive and end time is inclusive).
+		if !key.Bucket.Before(ts.GoTime()) {
+			continue
+		}
+
+		// TODO(dan): These files should be further subdivided for three
+		// reasons. 1) we could always gc anything we flush and later write a
+		// followup bucket subdivion if needed 2) very large bucket sizes could
+		// mean very large files, which are unwieldy once written 3) smooth
+		// and/or control memory usage of the sink.
+		filename := key.Filename()
+		if log.V(1) {
+			log.Info(ctx, "writing ", filename)
+		}
+		if err := s.writeFile(ctx, filename, file); err != nil {
+			return err
+		}
+
+		// If the bucket end is `<= ts`, we'll never see another row for this
+		// bucket and it can be cleaned up.
+		if end := key.Bucket.Add(s.bucketSize); ts.GoTime().After(end) {
```

Comment:
Do we have any assertions for this invariant (never seeing another key) being true up the stack? It just seems in this case, we actually risk overwriting files if we get an unacceptably earlier key.